packages/nuxt/src/components/plugins/loader.ts

```diff
@@ -14,103 +15,243 @@ interface LoaderOptions {
+const REPLACE_COMPONENT_TO_DIRECT_IMPORT_RE = /(?<=[ (])_?resolveComponent\(\s*["'](lazy-|Lazy(?=[A-Z]))?(Idle|Visible|idle-|visible-|Interaction|interaction-|MediaQuery|media-query-|If|if-|Never|never-|Time|time-)?([^'"]*)["'][^)]*\)/g
 export const LoaderPlugin = (options: LoaderOptions) => createUnplugin(() => {
   const exclude = options.transform?.exclude || []
   const include = options.transform?.include || []
   const nuxt = tryUseNuxt()
 
-  return {
-    name: 'nuxt:components-loader',
-    enforce: 'post',
-    transformInclude (id) {
-      if (exclude.some(pattern => pattern.test(id))) {
-        return false
-      }
-      if (include.some(pattern => pattern.test(id))) {
-        return true
-      }
-      return isVue(id, { type: ['template', 'script'] }) || !!id.match(SX_RE)
-    },
-    transform (code, id) {
-      const components = options.getComponents()
-
-      let num = 0
-      const imports = new Set<string>()
-      const map = new Map<Component, string>()
-      const s = new MagicString(code)
-
-      // replace `_resolveComponent("...")` to direct import
-      s.replace(REPLACE_COMPONENT_TO_DIRECT_IMPORT_RE, (full: string, lazy: string, name: string) => {
-        const component = findComponent(components, name, options.mode)
-        if (component) {
-          // TODO: refactor to nuxi
-          const internalInstall = ((component as any)._internal_install) as string
-          if (internalInstall && nuxt?.options.test === false) {
-            if (!nuxt.options.dev) {
-              const relativePath = relative(nuxt.options.rootDir, id)
-              throw new Error(`[nuxt] \`~/${relativePath}\` is using \`${component.pascalName}\` which requires \`${internalInstall}\``)
+  return [
+    {
+      name: 'nuxt:components-loader-pre',
+      enforce: 'pre',
+      transformInclude (id) {
+        if (exclude.some(pattern => pattern.test(id))) {
+          return false
+        }
+        if (include.some(pattern => pattern.test(id))) {
+          return true
+        }
+        return isVue(id, { type: ['template'] })
+      },
+      async transform (code) {
+        // change <LazyMyComponent hydrate-on-idle /> to <LazyIdleMyComponent hydrate-on-idle />
+        const { 0: template, index: offset = 0 } = code.match(TEMPLATE_RE) || {}
+        if (!template) { return }
+        if (!LAZY_HYDRATION_PROPS_RE.test(template)) {
+          return
+        }
+        const s = new MagicString(code)
+        try {
+          const ast = parse(template)
+          const components = options.getComponents()
+          await walk(ast, (node) => {
+            if (node.type !== 1 /* ELEMENT_NODE */) {
+              return
+            }
+            if (!/^(?:Lazy|lazy-)/.test(node.name)) {
+              return
+            }
+            const pascalName = pascalCase(node.name.slice(4))
+            if (!components.some(c => c.pascalName === pascalName)) {
+              // not auto-imported
```

Comment:
Does that meant that lazy hydration will only work for auto-imported components? 🤔

Reply:
I guess that should be documented then (or did I miss that part?)

Hmm, maybe analyzing the import and creating a wrapper for the imported component "on the fly"? 🤔

Reply:
I think that would be the right solution 👍